include/aws/crt/ImdsClient.h

```diff
@@ -0,0 +1,342 @@
+struct aws_imds_instance_info;
+
+namespace Aws
+{
+
+    namespace Crt
+    {
+
+        namespace Io
+        {
+            class ClientBootstrap;
+        }
+
+        namespace Auth
+        {
+            class Credentials;
+        }
+
+        namespace Imds
+        {
+
+            struct AWS_CRT_CPP_API ImdsClientConfig
+            {
+                ImdsClientConfig() : Bootstrap(nullptr) {}
+
+                /**
+                 * Connection bootstrap to use to create the http connection required to
+                 * query resource from the Ec2 instance metadata service
+                 */
+                Io::ClientBootstrap *Bootstrap;
+
+                /* Should add retry strategy support once that is available */
+            };
+
+            /**
+             * https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
+             */
+            struct AWS_CRT_CPP_API IamProfile
+            {
+                DateTime lastUpdated;
+                ByteCursor instanceProfileArn;
+                ByteCursor instanceProfileId;
+            };
+
+            /**
+             * Block of per-instance EC2-specific data
+             *
+             * https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-identity-documents.html
+             */
+            struct AWS_CRT_CPP_API InstanceInfo
+            {
+                /* an array of aws_byte_cursor */
+                Vector<ByteCursor> marketplaceProductCodes;
+                ByteCursor availabilityZone;
+                ByteCursor privateIp;
+                ByteCursor version;
+                ByteCursor instanceId;
+                /* an array of aws_byte_cursor */
+                Vector<ByteCursor> billingProducts;
+                ByteCursor instanceType;
+                ByteCursor accountId;
+                ByteCursor imageId;
+                DateTime pendingTime;
+                ByteCursor architecture;
+                ByteCursor kernelId;
+                ByteCursor ramdiskId;
+                ByteCursor region;
+            };
+
+            using OnResourceAcquired = std::function<void(const ByteBuf *resource, int errorCode, void *userData)>;
+            using OnVectorResourceAcquired =
+                std::function<void(const Vector<ByteCursor> &resource, int errorCode, void *userData)>;
+            using OnCredentialsAcquired =
+                std::function<void(const Auth::Credentials &credentials, int errorCode, void *userData)>;
+            using OnIamProfileAcquired =
+                std::function<void(const IamProfile &iamProfile, int errorCode, void *userData)>;
+            using OnInstanceInfoAcquired =
+                std::function<void(const InstanceInfo &instanceInfo, int errorCode, void *userData)>;
+
+            class AWS_CRT_CPP_API ImdsClient
+            {
+              public:
+                ImdsClient(aws_imds_client *client, Allocator *allocator = g_allocator) noexcept;
```

Comment:
Do we need this constructor?

Reply:
I did this following `Credentials`'s ctor. Shall we allow customer to initiate a Client using the raw client?

Reply:
Credentials has that constructor so that we can give something back to the user from a C callback.  With the imds client, a constructor from config is sufficient.